### ISSUE TITLE
Faster ifelse() in base R

### ISSUE THREAD

**Issue by bobaekang** [ICJIA/r-user-group#11]
## Summary
A faster base R alternative to `ifelse()`.

### Introducing `ifelse()`
`ifelse()` is a base R function that takes a logical `test` vector, a value if `test` is `TRUE` and a value if `test` is `FALSE`. It can be thought as "vectorized" if-else, which takes a vector as an input, tests each of its elements using some criteria and returns another vector whose elements are based on the test result for each element of the input vector.

Here is a simple example where `ifelse()` returns `"one"` if a vector element equals 1, and `"not one"` otherwise:
```r
x <- c(1, 2, 3, 1, 4)

ifelse(x == 1, "one", "not one")
#> [1] "one"     "not one" "not one" "one"     "not one"
```

Despite its convenience, `ifelse()` is known to be slow and is often responsible for performance bottlenecks. In the following sections, we will see if we can improve the situation by writing a custom function.

### A faster base R alternatives to `ifelse()`
A minimalist base R alternative to `ifelse()` looks as follows:
```r
fast_ifelse <- function(test, yes, no) {
  out <- rep(NA, length(test))
  out[test] <- yes
  out[!test] <- no
  
  out
}
```

`fast_ifelse()` first initializes the output vector, `out`,  which contains only `NA`s. The following two lines then replaces elements of `out` with the `yes` input where `test` is `TRUE`, i.e. `out[test]`, and with the `no` input where `test` is `FALSE`, i.e. `out[!test]`. Finally, the function returns `out`. 

How faster is `fast_ifelse()` than `ifelse()`? Let’s compare them using `microbenchmark::microbenchmark()`. The first benchmarking uses a numeric vector of 1 million integers at a random order. The `test` checks whether an element is smaller than 1000 or not. The second benchmarking involves a slightly more complicated `test` that checks which of 1 million random alphabet letters is "a", "b", or "c".

So here we go:
```r
set.seed(1234)
x <- sample(1:1e6, 1e6)
y <- sample(letters, 1e6, replace = TRUE)

# install.packages("microbenchmark")

microbenchmark::microbenchmark(
  ifelse(x < 1000, "small", "big"),
  fast_ifelse(x < 1000, "small", "big")
)
#> Unit: milliseconds
#>                                   expr      min       lq      mean
#>       ifelse(x < 1000, "small", "big") 470.7486 495.8225 544.12860
#>  fast_ifelse(x < 1000, "small", "big")  37.3075  47.8843  56.47106
#>     median        uq      max neval
#>  541.35680 585.00945 690.9175   100
#>   54.36975  61.22485 141.5901   100

microbenchmark::microbenchmark(
  ifelse(y %in% c("a", "b", "c"), "abc", "other"),
  fast_ifelse(y %in% c("a", "b", "c"), "abc", "other")
)
#> Unit: milliseconds
#>                                                  expr      min        lq
#>       ifelse(y %in% c("a", "b", "c"), "abc", "other") 509.3277 534.94605
#>  fast_ifelse(y %in% c("a", "b", "c"), "abc", "other")  63.2084  78.20995
#>       mean   median       uq       max neval
#>  614.76175 588.3243 640.7817 1189.6410   100
#>   90.32842  83.6735  90.1476  248.3969   100
```
The result show that `fast_ifelse()` is **~10x faster**:zap::zap: than that of the original `ifelse()` in the first comparison! With a slightly more complicated `test`, `fast_ifelse()` is still **~7x faster**:zap:.

### Better safe than sorry
In addition to its poor performance, `ifelse()` is known to surprise its users with unexpected results as it strips away the attributes of `yes` and `no` input objects.

See the following example:
```r
x <- c(2008, 2008, 2018, 2018)
yes <- as.Date("2008-01-01")
no <- as.Date("2018-01-01")

# result is numeric, since Date is built on double (numeric)
ifelse(x == 2008, yes, no)
#> [1] 13879 13879 17532 17532

# ...oh, no!
fast_ifelse(x == 2008, yes, no)
#> [1] 13879 13879 17532 17532
```

We find that, even though we used `Date` values for `yes` and `no`, the `ifelse()` output is a `numeric` class object. This is because a `Date` object is built on top of the `double`(`numeric`) type vector with an added `class` attribute of `Date`, which is then stripped away by the internal operations of `ifelse()`. In the same example, we find that `fast_ifelse()`, too, fails to preserve the input attributes.

Fortunately, this can be easily remedied with only two more lines of code:
```r
fast_ifelse2 <- function(test, yes, no) {
  stopifnot(identical(class(yes), class(no)))
 
  out <- rep(NA, length(test))
  out[test] <- yes
  out[!test] <- no
  class(out) <- class(yes) 
  
  out
}

# result is Date
fast_ifelse2(x == 2008, yes, no)
#> [1] "2008-01-01" "2008-01-01" "2018-01-01" "2018-01-01"
```

The first line of the body of `fast_ifelse2()` checks whether the `yes` and `no` inputs are of the same class--and throws an error if not. This is to ensure that the final result is also of the same class without relying on some implicit mechanisms. Then, after filling in `out` with  `yes` and `no` values, `fast_ifelse2()`  assigns to `out` the class of `yes`, which has already been verified to be equal to that of `no`. Finally, `fast_ifelse2()` returns `out`, now of the intented class.

Luckily, the work of checking and assigning class has only a minimal impact on performance. This means that we now have an alternative to `ifelse()` that is not only faster but also safer:rescue_worker_helmet:.

### vs. `dplyr::if_else()`
`dplyr`, a popular R package for data manipulation, offers its version of `ifelse()`, i.e. `dplyr::if_else()`, whose performance is superior to that of the original. In fact, the custom solution discussed so far is partly inspired by `dplyr::if_else()`. See `?dplyr::if_else` or [this reference page](https://dplyr.tidyverse.org/reference/if_else.html) for more details.

While `faster_ifelse2()` comes with a benefit of not relying on any third-party package, the widespread use of `dplyr` in data manipulation tasks offers a strong case for using `dplyr::if_else()` instead of the custom solution; that is, *if* the performance of a custom solution is at least practically indistinguishable from that of `dplyr::if_else()`.

This only means one thing: let's compare:
```r
microbenchmark::microbenchmark(
  fast_ifelse2(x < 1000, "small", "large"),
  dplyr::if_else(x < 1000, "small", "large")
)
#> Unit: milliseconds
#>                                        expr     min       lq      mean
#>    fast_ifelse2(x < 1000, "small", "large") 39.0669 51.37315  65.61674
#>  dplyr::if_else(x < 1000, "small", "large") 79.3622 97.22840 133.66975
#>     median       uq      max neval
#>   58.06705  65.7427 194.6295   100
#>  105.53400 175.0614 543.1557   100

microbenchmark::microbenchmark(
  fast_ifelse2(y %in% c("a", "b", "c"), "abc", "other"),
  dplyr::if_else(y %in% c("a", "b", "c"), "abc", "other")
)
#> Unit: milliseconds
#>                                                     expr      min       lq
#>    fast_ifelse2(y %in% c("a", "b", "c"), "abc", "other")  68.0187  82.0125
#>  dplyr::if_else(y %in% c("a", "b", "c"), "abc", "other") 109.1392 126.3121
#>       mean    median        uq      max neval
#>   93.09773  88.24975  93.72515 194.5924   100
#>  153.45127 134.63525 185.43690 334.4240   100
```

Hey, the base R solution is still **~2x faster** than `dplyr`'s! A significant improvement for sure:heart_eyes:.

`dplyr::if_else()`  offers a somewhat greater degree of flexibility (e.g. an ability to handle missing values) and is, arguably, more robust. Nonetheless, a nontrivial performance gain from using `fast_ifelse2()` and its zero dependency, together, provide a compelling case for choosing `fast_ifelse2()` over `dplyr::if_else()`--perhaps even when `dplyr` is already loaded.

### TL;DR
* The base R `ifelse()` is a "vectorized" if-else that is useful but too slow.
* `fast_ifelse()` is a minimalist alternative that can serve as a drop-in replacement for `ifelse()`.
* `fast_ifelse2()` also preserves the class of the input values, thus improving the predictabilty.
* `dplyr::if_else()` is yet another alternative that is faster and more strict than `ifelse()`.
* `fast_ifelse2()` can be significantly faster than both `ifelse()` and `dplyr::if_else()`.
